backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/ChangeLogUtils.java

```diff
@@ -14,22 +14,58 @@ private ChangeLogUtils() {
 
   public static String buildProcessAuditFunction(String schemaName, String tableName) {
     return """
-            CREATE OR REPLACE FUNCTION "%1$s"."process_%3$s_audit"() RETURNS TRIGGER AS $%3$s_audit$
-                   BEGIN
-                       IF (TG_OP = 'DELETE') THEN
-                           INSERT INTO "%1$s".mg_changelog
-                           SELECT 'D', now(), user, TG_TABLE_NAME, row_to_json(OLD.*), row_to_json(NEW.*);
-                       ELSIF (TG_OP = 'UPDATE') THEN
-                           INSERT INTO "%1$s".mg_changelog
-                           SELECT 'U', now(), user, TG_TABLE_NAME, row_to_json(OLD.*), row_to_json(NEW.*);
-                       ELSIF (TG_OP = 'INSERT') THEN
-                           INSERT INTO "%1$s".mg_changelog
-                           SELECT 'I', now(), user, TG_TABLE_NAME, row_to_json(OLD.*), row_to_json(NEW.*);
-                       END IF;
-                       RETURN NULL; -- result is ignored since this is an AFTER trigger
-                   END;
-                   $%3$s_audit$ LANGUAGE plpgsql;
-                """
+        CREATE OR REPLACE FUNCTION "%1$s"."process_%3$s_audit"()
+        RETURNS TRIGGER AS $%3$s_audit$
+        DECLARE
+            old_row JSONB;
+            new_row JSONB;
+            col_name TEXT;
+            old_value TEXT;
+            new_value TEXT;
+        BEGIN
+            -- Initialize empty JSONB objects
+            old_row := '{}'::JSONB;
+            new_row := '{}'::JSONB;
+
+            -- Loop through each column in the OLD record
+            FOR col_name IN SELECT column_name
+                             FROM information_schema.columns
+                             WHERE table_name = TG_TABLE_NAME
+            LOOP
+                -- Skip columns that end with '_contents' or '_TEXT_SEARCH_COLUMN'
```

Comment:
yes, that is a major pitfall that I also noticed. There is a better but more complex way to do it, which is that we use the metadata to generic the exact trigger we want. Then we can check for columnType=FILE. I can do that, because we can be quite sure that _content will exist. (With hindsight, the column should have been named something like _mg_content  or another way to clearly indicate internal use)

Reply:
yes, agree that we have time to fix this properly